### PR TITLE
DE-20702 New delay rule

### PR DESCRIPTION
**Pull request by TomasHodan01**

### CODE REVIEW

File: src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private function validateDefinition(array $delayDefinition): void
+    {
+        $attemptsDefinition = array_keys($delayDefinition);
+        $attemptsDefinitionCount = count($attemptsDefinition);
+
+        if (!in_array(0, $attemptsDefinition, true)) {
+            throw DelayRuleException::byMissingDefinitionForZeroAttempts();
+        }
+
+        for ($i = 1; $i < $attemptsDefinitionCount; $i++) {
```
Comment (davidpilny) [inline src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php:81]: Not really sure what we validate here. So $i is attempsCount? and for each attemps count there should be a definition?

File: src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        $this->linearDelayDefinition = $linearDelayDefinition;
+    }
+
+
+    public function getDelay(JobInterface $job, Throwable $exception): int
+    {
+        $currentJobAttempts = $job->getAttempts();
+        $delay = 0;
+
+        foreach ($this->constantDelayDefinition as $attempts => $delayInSeconds) {
```
Comment (davidpilny) [inline src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php:53]: Why is there a combination of constant and linear? What is the idea?

**Review by davidpilny — COMMENTED** [review]
I dont see where it will be used so not sure what we want to achieve.

File: src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php (hunk 2 above)
Comment (lukasHod) [inline src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php:53]: $attempts is little confusing naming. Maybe ruleDelayAttempt?

File: src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        $attemptsDefinition = array_keys($delayDefinition);
+        $attemptsDefinitionCount = count($attemptsDefinition);
+
+        if (!in_array(0, $attemptsDefinition, true)) {
+            throw DelayRuleException::byMissingDefinitionForZeroAttempts();
+        }
+
+        for ($i = 1; $i < $attemptsDefinitionCount; $i++) {
+            if ($attemptsDefinition[$i - 1] < $attemptsDefinition[$i]) {
```
Comment (davidpilny) [inline src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php:67]: $i -> $attemptsCount? :angel:

Comment (davidpilny) [reply src/Jobs/FailResolving/DelayRules/DefinedDelayRule.php:67]: or limitAttempts?

**Review by lukasHod — APPROVED** [review]

**Review by milan-zelenka — APPROVED** [review]

**Pull request merged**